### PR TITLE
HPCC-27267 LDAPEnvironment to create LDAPAdmin and HPCCAdmin accounts

### PR DESCRIPTION
**Pull request by RussWhitehead**
Create an HPCC Admin account in the configured UserOU, create the k8s secret
for that user, add that user to the HPCCAdmin group, and grant that group
access to ECLWatch (SmcAccess)
Create an LDAP Admin account in the configured UserOU, create the k8s secret
for that user, and add LDAP R/W permissions for that user to access within
the environment tree (not outside).
For both users, generate the HELM that the caller can copy/paste into their
Helm Chart override file

Signed-off-by: Russ Whitehead <russwhitehead@yahoo.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [x] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [x] Performance
  - [x] Security
  - [ ] Thread-safety
  - [x] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [x] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [x] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

**Comment by RussWhitehead** [comment]
@afishbeck Please review
I tested by generating an environment, copying the generated yaml into an override file, and started ESP using the new environment. Verified generated LDAPAdmin only has access to private environment scope. Logged into ECLWatch with 
generated HPCCAdmin account, verified it had SmcAccess, navigated through ECLWatch, went into security and created/updated users/groups/permissions

File: esp/services/ldapenvironment/ldapenvironmentService.cpp [hunk 1]
```diff
@@ -159,77 +207,147 @@ bool CldapenvironmentEx::onLDAPCreateEnvironment(IEspContext &context, IEspLDAPC
...
-            VStringBuffer cmdLine("%s%s", cmdLineSafe.str(), respHPCCAdminPwd.str());
-            try
+            //Create the HPCCAdmin user
             {
-                DBGLOG("\nExecuting '%s'\n", cmdLineSafe.str());
-                long unsigned int runcode;
-                bool success = invoke_program(cmdLine.str(), runcode) && (runcode==0);
-                if (!success)
-                    notes.appendf("\nError executing '%s', please run manually before executing helm install", cmdLineSafe.str());
+            Owned<ISecUser> user;
```
Comment (afishbeck) [inline esp/services/ldapenvironment/ldapenvironmentService.cpp:242]: With quite a few of these changes you seem to have lost one level of indentation.  We typically indent inside the curly brace, which the old code used to do.

Comment (afishbeck) [reply esp/services/ldapenvironment/ldapenvironmentService.cpp:242]: In cases where you are setting the owned variable right away it's easier to read if you assign directly:

`Owned<ISecUser> user = secmgr->createUser(respHPCCAdminUser.str());`

File: esp/services/ldapenvironment/ldapenvironmentService.cpp [hunk 2]
```diff
@@ -159,77 +207,147 @@ bool CldapenvironmentEx::onLDAPCreateEnvironment(IEspContext &context, IEspLDAPC
...
+        // Attempt to create the Kubernetes secret for that user
+        //----------------------------------
+        VStringBuffer respLDAPAdminUser("adminLDAP_%s", req.getEnvName());
+        StringBuffer  respLDAPAdminPwd;
+        generatePassword(respLDAPAdminPwd, 10);//jutil.hpp
+
+        if (req.getCreateLDAPEnvironment())
+        {
+            //Create the user
+            Owned<ISecUser> user;
```
Comment (afishbeck) [inline esp/services/ldapenvironment/ldapenvironmentService.cpp:297]: Again might as well assign directly, it's easier to read.

File: system/jlib/jutil.cpp [hunk 3]
```diff
@@ -3347,25 +3347,27 @@ const char * generatePassword(StringBuffer &pwd, int pwdLen)
 
     //Ensures each character group used at least once
     srand(time(0));
-    pwd.append(alphaUC[rand() % sizeof(alphaUC)]);
-    pwd.append(alphaLC[rand() % sizeof(alphaLC)]);
-    pwd.append(numeric[rand() % sizeof(numeric)]);
-    pwd.append(symbol[rand() % sizeof(symbol)]);
+    pwd.append(alphaUC[rand() % (sizeof(alphaUC) - 1)]);
```
Comment (afishbeck) [inline system/jlib/jutil.cpp:3350]: Will this ever output a 'Z'?

File: system/jlib/jutil.cpp [hunk 4]
```diff
@@ -3347,25 +3347,27 @@ const char * generatePassword(StringBuffer &pwd, int pwdLen)
...
+            pwd.append(alphaLC[rand() % (sizeof(alphaLC) - 1)]);  break;
         case 2:
-            pwd.append(numeric[rand() % sizeof(numeric)]);  break;
+            pwd.append(numeric[rand() % (sizeof(numeric) - 1)]);  break;
         case 3:
-            pwd.append(symbol[rand() % sizeof(symbol)]);  break;
+            pwd.append(symbol[rand() % (sizeof(symbol) - 1)]);  break;
         }
     }
+    if (pwd.length() != (size32_t)pwdLen)
```
Comment (afishbeck) [inline system/jlib/jutil.cpp:3369]: Would this ever happen?

File: system/security/LdapSecurity/ldapconnection.cpp [hunk 5]
```diff
@@ -4185,13 +4198,17 @@ class CLdapClient : implements ILdapClient, public CInterface
...
+            //if not already fully qualified
+            if (0 == strstr(action.m_account_name, "dc=") &&
+                0 == strstr(action.m_account_name, "ou="))
+            {
+                StringBuffer act_dn;
+                if(action.m_account_type == GROUP_ACT)
+                    getGroupDN(action.m_account_name.str(), act_dn);
+                else
+                    getUserDN(action.m_account_name.str(), act_dn);
+                action.m_account_name.clear().append(act_dn.str());
```
Comment (afishbeck) [inline system/security/LdapSecurity/ldapconnection.cpp:4210]: Just a note that `s.set(str)` is equivalent to `s.clear().append(str);` so you could do:

`action.m_account_name.set(act_dn);`

File: system/security/LdapSecurity/ldapsecurity.cpp [hunk 6]
```diff
@@ -1630,6 +1635,11 @@ const bool CLdapSecManager::organizationalUnitExists(const char * ou) const
     return m_ldap_client->organizationalUnitExists(ou);
 }
 
+bool CLdapSecManager::addUser(ISecUser & user, const char* basedn)
+{
+	return m_ldap_client->addUser(user, basedn);
```
Comment (afishbeck) [inline system/security/LdapSecurity/ldapsecurity.cpp:1640]: Indented too far.

**Review by afishbeck — COMMENTED** [review]
Much of the formatting isn't done in typical HPCC style.  Unfortunately I'm focusing on the c++ because I'm not that knowledgeable about the implication of how the actual admin permissions should be set to do what we want. 
Ideally these admin users can't change anything outside of the OUs for this environment.
As we've discussed offline there should be another JIRA for figuring out how to deal with changing of passwords.

File: esp/services/ldapenvironment/ldapenvironmentService.cpp [hunk 7]
```diff
@@ -108,6 +108,54 @@ const char * CldapenvironmentEx::formatOUname(StringBuffer &ou, const char * env
...
+    action.m_basedn = baseDN;
+    action.m_rname =  rName;
+    action.m_rtype = RT_DEFAULT;
+    action.m_account_name = userFQDN;//fully qualified user DN
+    action.m_account_type = USER_ACT;
+    action.m_allows = allows;
+    action.m_denies = denies;
+    DBGLOG("Setting (%d,%d) permissions for rName %s, baseDN %s, user %s", (int)allows, (int)denies, rName.str(), baseDN.str(), userFQDN);
+    bool ok;
+    try { ok = secmgr->changePermission(action); }
```
[the remark this answers is not in the document]
Comment (RussWhitehead) [reply esp/services/ldapenvironment/ldapenvironmentService.cpp:135]: Fixed all instances

File: system/jlib/jutil.cpp (hunk 3 above)
Comment (RussWhitehead) [reply system/jlib/jutil.cpp:3350]: Yes, before this fix it was outputting NULL, resulting in a truncated password

File: system/jlib/jutil.cpp (hunk 4 above)
Comment (RussWhitehead) [reply system/jlib/jutil.cpp:3369]: It was before fixing above bug, this is a sanity check. I'll make it a IFDEF_DEBUG

**Comment by RussWhitehead** [comment]
@afishbeck The LDAP admin is granted RW access to the environment OU, and everything under it. It only gets READ access to outside OUs, which could include "shared/global" ou and hand crafted ones that are not part of the environment
The password-change Jira is https://track.hpccsystems.com/browse/HPCC-27268

**Review by afishbeck — APPROVED** [review]
@RussWhitehead looks good.

File: esp/scm/ldapenvironment.ecm [hunk 8]
```diff
@@ -44,7 +44,7 @@ ESPenum OUMode : int
 ESPrequest LDAPCreateEnvironmentRequest
 {
     bool                                              CreateLDAPEnvironment(false);
-    bool                                              CreateK8sSecret(false);
+    bool                                              CreateK8sSecrets(false);
```
Comment (richardkchapman) [inline esp/scm/ldapenvironment.ecm:47]: I'm not sure exactly how ECM files work, but I thought there was some versioning mechanism we used when changing interfaces. This feels like it's going to break anyone using the prior interface.

**Review by richardkchapman — CHANGES_REQUESTED** [review]

File: esp/scm/ldapenvironment.ecm (hunk 8 above)
Comment (afishbeck) [reply esp/scm/ldapenvironment.ecm:47]: That's true, although this service is still brand new and in the proof-of-concept stage... so not sure it's necessary.

But changing the name of an input and adding fields to the output would normally both be versioned.

Comment (RussWhitehead) [reply esp/scm/ldapenvironment.ecm:47]: @richardkchapman As Tony said, this is brand new code that nobody has looked at yet.

**Comment by RussWhitehead** [comment]
@richardkchapman Please merge

**Review by richardkchapman — APPROVED** [review]

**Pull request merged**